proposals/0000-updated-partial-type-sigs.rst

```diff
@@ -0,0 +1,726 @@
+|                            |f = _``.                |                              |
+|                            |                        |                              |
+|                            |Covers 1AB/2A/3C/4A/5B  |Covers 1AB/2C/3AB/4B/5B       |
+|                            |                        |                              |
+|                            |                        |                              |
++----------------------------+------------------------+------------------------------+
+						  
+
+Effect and Interactions
+-----------------------
+* All positions marked **Not allowed** (outside of patterns 5A)
+  in the Motivation are now allowed.
+
+* The new design allows the user to control whether they want an elision or a named wildcard, using
+  a convenient naming convention.
+
+* The new design gives users fine control over generalization, through the use of ``%Partial`` to
+  suppress kind generalization and the use of ``forall a b c _.`` to explicitly enable type
+  generalization.
+
+* Visible kind application now fits in nicely. Users can control whether they want elisions
+  or wildcards.
+
+* Partial type signatures have become louder, through the addition of ``%Partial``. This makes it
+  more sensible to keep partial type signatures in released code. The new syntax also
+  allows users to write elisions in type signatures without causing the signature to
+  become partial.
+
+* Wildcards might look like definitions in lens-heavy code; that is, a misspelled ``_field``
+  would now be a wildcard. I don't think the ensuing error message would pose a challenge
+  to a programmer in figuring out what happened. We can continue to suggest possible misspellings
+  in such error messages.
+
+* The specification above describes behavior outside of patterns. No change is made to
+  the way patterns behave. Note that type signatures in patterns are not patterns.
+  
+Costs and Drawbacks
+-------------------
+* Partial type signatures have proved hard to implement and with many corner cases. The new
+  design seems no simpler nor more complicated than the current, but it will take a fair amount
+  of work to re-engineer.
+
+* The new design does not adequately treat patterns. It is conceivable that a user would want
+  a wildcard (with diagnostic information) in a pattern, and this is no more achievable with this
+  proposal than it was previously.
+
+* This proposal is not backward compatible, as it requires the ``%Partial`` modifier for partial
+  type signatures.
+
+* This proposal warns on the Haskell98 program ::
+
+    id :: _w -> _w
+    id x = x
+
+  Thus, this standards-conforming program would now cause GHC to bleat (but still accept,
+  with its original meaning).
+  
+* Elisions in types (``const :: a -> _ -> a``) seem less useful than other aspects of this
+  proposal, and yet they occupy prime syntactic real estate. Is there a better design around
+  this issue?
+  
+Alternatives
```

Comment:
After reading through this proposal again (after four years!!), I think there may be a simpler approach that would capture most of the value.

I think there are two UX modalities to consider here:

1. _"I don't **care** what goes here."_ This covers existing pattern wildcards and type expression elisions, both of which use `_` and seem to work well. There should be no warning for this, the programmer is explicitly saying they do not care. I think it also covers some use cases for partial type signatures, though sometimes the programmer does care, they just need GHC's help figuring it out.
2. _"I don't **know** what goes here."_ This covers expression holes and the other cases of partial type signatures. These both also use `_` which is confusing! The programmer is asking GHC a question here, not expressing ambivalence. We should have different syntax for this modality; I propose `?` (or `?x` for named queries) as a strawman. This is a developer tool, so the semantics should probably be that GHC prints out the inferred answer as a warning, but proceeds with compilation (injecting bottom values as necessary).

@goldfirere what do you think of this split?

Reply:
Very happy with that. Other than your proposed concrete syntax, I don't really see the difference between what you've written and parts 1 and 2 of my proposed change. Your suggestion does not cover when we should generalize type signatures, which is a subtle but important point in practice.

Reply:
@goldfirere I think there are some differences, let me try to reproduce your summary table with my thinking.

  | elision (spelled `_`) | hole (spelled `?`)
-- | -- | --
in a type signature | GHC uses unification to fill in elision. If unification does not find a value for the elision, generalize. | GHC treats the signature as if it were partial and uses the definition of the identifier to solve for the hole. Issues an error with both the value GHC has discovered for the wildcard and its kind.
in a visible type application | same as above, but no generalization. If GHC is unable to figure out what an elision should be, error. | same behavior as an elision, but issues an error with the inferred type
in another type | same as in a type signature | issue an error with the kind of the hole and any information GHC can figure out about the content of the hole.
in an expression | GHC replaces the underscore with `error "elision at..."`. | Same behavior as today's holes: issue an error with the hole's type and suggestions for what it might be filled in with.

Note that elisions do not abort compilation unless GHC truly can't figure out what to do, whereas holes always abort compilation with suggestions for filling the hole. Also note that we don't need a `%Partial` modifier any more, a partial type signature is one that contains a hole `?`.

Finally, I think both elisions and holes could be extended with named variants `_x` and `?x` to allow sharing.